app/replica.go

```diff
@@ -84,29 +84,46 @@ func CheckReplicaState(frontendIP string, replicaIP string) (string, error) {
 }
 
 func AutoConfigureReplica(s *replica.Server, frontendIP string, address string, replicaType string) {
-checkagain:
-	state, err := CheckReplicaState(frontendIP, address)
-	logrus.Infof("Replicastate: %v err:%v", state, err)
-	if err == nil && (state == "" || state == "ERR") {
+	var (
+		err   error
+		state string
+	)
+	for {
+		state, err = CheckReplicaState(frontendIP, address)
+		logrus.Infof("Replicastate: %v err:%v", state, err)
+
+		if err != nil {
+			logrus.Infof("checkReplicaState failed, err:%v retry after 5sec..", err)
+			time.Sleep(5 * time.Second)
+			continue
+		}
+		if state == "WO" || state == "ERR" {
+			if err = AutoRmReplica(frontendIP, address); err != nil {
+				logrus.Warning("AutoRmReplica failed, retry after 5 second")
+				time.Sleep(5 * time.Second)
+				continue
+			}
+		} else if state == "RW" {
+			time.Sleep(5 * time.Second)
+			continue
+		}
 		s.Close(false)
-	} else {
-		time.Sleep(5 * time.Second)
-		goto checkagain
-	}
-	AutoRmReplica(frontendIP, address)
-	AutoAddReplica(s, frontendIP, address, replicaType)
-	logrus.Infof("Waiting on MonitorChannel")
-	select {
-	case <-s.MonitorChannel:
+		if err = AutoAddReplica(s, frontendIP, address, replicaType); err != nil {
+			logrus.Warning("AutoAddReplica failed, retry after 5 second")
+			time.Sleep(5 * time.Second)
+			continue
+		}
+
+		logrus.Infof("Waiting on MonitorChannel")
```

Comment:
We are waiting on monitorchannel so that RPC will write onto this? or any other go routine does this?

Reply:
Correct, rpc and s.Close both can write on it.